### PR TITLE
Fix version helper classes

### PR DESCRIPTION
**Pull request by gcx-seb**
I think the `Version` class has some issues which I tried to fix. I have added some test cases which fail without these changes. Feel free to have a look.

Also I think `Version.V1_0`, `LwM2mVersion.V1_0` and `LwM2mVersion.V1_1` should probably be `final`… so I changed that, too. Happy to hear your opinions.

### CODE REVIEW

**Comment by sbernard31** [comment]
That totally makes sense to me.

File: leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java [hunk 1]
```diff
@@ -157,8 +157,8 @@ public static String validate(String version) {
         public int hashCode() {
             final int prime = 31;
             int result = 1;
-            result = prime * result + ((major == null) ? 0 : major.hashCode());
-            result = prime * result + ((minor == null) ? 0 : minor.hashCode());
+            result = prime * result + (int) major;
+            result = prime * result + (int) minor;
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java:161]: When I generate `hashcode()` with **eclipse**. 
There is no cast here. 

Code looks like : 
```java
result = prime * result + major;
result = prime * result + minor;
```

Any reason to use cast here ?

File: leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java [hunk 2]
```diff
@@ -91,35 +91,35 @@ public boolean equals(Object obj) {
...
 
         public Version(int major, int minor) {
-            this.major = (short) major;
-            this.minor = (short) minor;
+            this((short) major, (short) minor);
+            if (this.major != major) {
+                throw new ArithmeticException("short overflow");
+            }
+            if (this.minor != minor) {
+                throw new ArithmeticException("short overflow");
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java:106]: Maybe the error message could looks like the one in `validate()` method ? 
maybe something like : 
- `"version (%d.%d) major part (%d) is not a valid short"`
- `"version (%d.%d) minor part (%d) is not a valid short"`

File: leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java [hunk 3]
```diff
@@ -91,35 +91,35 @@ public boolean equals(Object obj) {
...
+            }
         }
 
-        public Version(Short major, Short minor) {
+        public Version(short major, short minor) {
             this.major = major;
             this.minor = minor;
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java:112]: should we add test to check that short is not negative ?

- `"version (%d.%d) major part %d must not be negative"`
- `"version (%d.%d) major part %d must not be negative"`

There is an issue with that, we will not be able to call `this((short) major, (short) minor);` in constructor above because error message could be wrong. I guess the code should be adapted a little more than just adding the check.

Any opinion about this ?

File: leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java (hunk 1 above)
Comment (gcx-seb) [reply leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java]: No specific reason, this was generated by IntelliJ :see_no_evil::slightly_smiling_face:

File: leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java (hunk 3 above)
Comment (gcx-seb) [reply leshan-core/src/main/java/org/eclipse/leshan/core/LwM2m.java]: Good point. I've added the negativity check there and refactored the int-to-short casting to include the correct error message.

**Review by sbernard31 — COMMENTED** [review]
Except this last question. 
I think we are good.

**Comment by sbernard31** [comment]
About **github-actions** comments, if you don't want to deal with it, it's not a big deal. I can do it myself :slightly_smiling_face:

File: leshan-core/src/test/java/org/eclipse/leshan/core/VersionTest.java [hunk 4]
```diff
@@ -35,9 +39,66 @@ public void is_supported_tests() {
     @Test
     public void compare_tests() {
         assertTrue(new Version("1.0").compareTo(new Version("1.2")) < 0);
+        assertTrue(new Version("1.0").olderThan(new Version("1.2")));
+        assertNotEquals(new Version("1.0").hashCode(), new Version("1.2").hashCode());
```
[the remark this answers is not in the document]
Comment (gcx-seb) [reply leshan-core/src/test/java/org/eclipse/leshan/core/VersionTest.java]: It was intended, so that the `hashCode()` generation is also verified. But I agree this became quite chaotic… I just cleaned that up (and also fixed the formatting issues hopefully :slightly_smiling_face:)

[the remark this answers is not in the document]
Comment (sbernard31) [reply leshan-core/src/test/java/org/eclipse/leshan/core/VersionTest.java]: If you want to check `equals`/`hashcode`, probably better to use `EqualsVerifier`. 

Replacing you code by  : 
```java
@Test
public void assertEqualsHashcode() {
    // TODO we should not use EqualsVerifier.simple()
    // but implement a right hashcode/equals way
    // see : https://github.com/eclipse-leshan/leshan/issues/1504
    EqualsVerifier.simple().forClass(Version.class).verify();
}
```
(If you are curious about that you could look at : #1504)

I will do this last little change and then integrate this PR in `master`.

**Comment by sbernard31** [comment]
This now integrated in `master` (squashed in 1 commit cb204b26adc97a3f290ea054ecd86799c47b3225)
This will be available in next release (2.0.0-M14, see #1531)

@gcx-seb Thx a lot for you contribution :pray:

**Comment by sbernard31** [comment]
@gcx-seb, could I ask how you are using Leshan ?

**Comment by gcx-seb** [comment]
> could I ask how you are using Leshan ?

Sure, nothing too fancy though… mainly just using `leshan-server-cf` for building LwM2M servers :slightly_smiling_face:

**Comment by sbernard31** [comment]
would you like to be added to Leshan adopters ? 
 - https://github.com/eclipse-leshan/leshan/issues/830
 - https://eclipse.dev/leshan/#who%20is%20using

**Comment by sbernard31** [comment]
Implementing #1535, I found that `String.split `default behavior is a bit strange and so discover more issue about `Version` class that I fixed in 2efb9240ecf352567db4b26e47e1a9e98e472109

**Comment by sbernard31** [comment]
(I forgot to close this issue :slightly_smiling_face:)

**Comment by gcx-seb** [comment]
> would you like to be added to Leshan adopters ?

Yes, we'd love to :heart: I'll post in #830 :+1:

**Comment by sbernard31** [comment]
@gcx-seb if you are OK I will create the PR to add you to : https://iot.eclipse.org/adopters/#iot.leshan

**Comment by gcx-seb** [comment]
Just looking over https://iot.eclipse.org/adopters/#iot.leshan, we're also using Californium, hawkBit, Mosquitto and Paho.

**Comment by sbernard31** [comment]
OK I will add it too. I let you know when the PR is ready (maybe not today)

**Comment by sbernard31** [comment]
I created the corresponding PR : https://gitlab.eclipse.org/eclipsefdn/it/api/eclipsefdn-project-adopters/-/merge_requests/241

**Comment by sbernard31** [comment]
(This is deployed : https://iot.eclipse.org/adopters/#iot.leshan)
